samcli/lib/telemetry/cicd.py

```diff
@@ -0,0 +1,120 @@
+
+    to handle both "CodeShip" and "codeship," here the string is converted to lower case first.
+
+    Parameters
+    ----------
+    environ
+
+    Returns
+    -------
+    bool
+        whether the env is CodeShip
+    """
+    ci_name: str = environ.get("CI_NAME", "").lower()
+    return ci_name == "codeship"
+
+
+_ENV_VAR_OR_CALLABLE_BY_PLATFORM: Dict[CICDPlatform, Union[str, Callable[[Mapping], bool]]] = {
+    # https://www.jenkins.io/doc/book/pipeline/jenkinsfile/#using-environment-variables
+    CICDPlatform.Jenkins: "JENKINS_URL",
+    # https://docs.gitlab.com/ee/ci/variables/predefined_variables.html
+    CICDPlatform.GitLab: "GITLAB_CI",
+    # https://docs.github.com/en/actions/reference/environment-variables
+    CICDPlatform.GitHubAction: "GITHUB_ACTION",
+    # https://docs.travis-ci.com/user/environment-variables/
+    CICDPlatform.TravisCI: "TRAVIS",
+    # https://circleci.com/docs/2.0/env-vars/
+    CICDPlatform.CircleCI: "CIRCLECI",
+    # https://docs.aws.amazon.com/codebuild/latest/userguide/build-env-ref-env-vars.html
+    CICDPlatform.AWSCodeBuild: "CODEBUILD_BUILD_ID",
+    # https://www.jetbrains.com/help/teamcity/predefined-build-parameters.html
+    CICDPlatform.TeamCity: "CODEBUILD_BUILD_ID",
+    # https://confluence.atlassian.com/bamboo/bamboo-variables-289277087.html
+    CICDPlatform.Bamboo: "bamboo_buildNumber",
+    # https://buddy.works/docs/pipelines/environment-variables
+    CICDPlatform.Buddy: "BUDDY",
+    CICDPlatform.CodeShip: _is_codeship,
+    # https://docs.semaphoreci.com/ci-cd-environment/environment-variables/
+    CICDPlatform.Semaphore: "SEMAPHORE",
+    # https://www.appveyor.com/docs/environment-variables/
+    CICDPlatform.Appveyor: "APPVEYOR",
+    CICDPlatform.Unknown: "CI",
+}
+
+
+def _is_cicd_platform(cicd_platform: CICDPlatform, environ: Mapping) -> bool:
+    """
+    Check whether sam-cli run in a particular CI/CD platform based on certain environment variables.
+
+    Parameters
+    ----------
+    cicd_platform
+        an enum CICDPlatform object indicating which CI/CD platform  to check against.
+    environ
+        the mapping to look for environment variables, for example, os.environ.
+
+    Returns
+    -------
+    bool
+        A boolean indicating whether there are environment variables matching the cicd_platform.
+    """
+    env_var_or_callable = _ENV_VAR_OR_CALLABLE_BY_PLATFORM[cicd_platform]
+    if isinstance(env_var_or_callable, str):
+        return env_var_or_callable in environ
+
+    # it is a callable, use the return value
+    return env_var_or_callable(environ)
+
+
+# detect CI/CD platform. Here we only calculate this value once so platform() can return this value
+# without re-calculating because CICD platform won't change once sam-cli starts.
+try:
+    _cicd_platform: Optional[CICDPlatform] = next(
+        cicd_platform for cicd_platform in CICDPlatform if _is_cicd_platform(cicd_platform, os.environ)
```

Comment:
Any order consideration in it? For example, if one env var appears in two platform. It will be fall in the first platform.

Reply:
And related to this, `dict` order isn't guaranteed on Python 3.6, so we shouldn't depend on it.

> Changed in version 3.7: Dictionary order is guaranteed to be insertion order. This behavior was an implementation detail of CPython from 3.6.

Reply:
yes, the environment variables are very platform-specific if you look at them. Except for the last one `CI`, which is placed at the bottom as a "catch-all"